### PR TITLE
Angular

### PR DESCRIPTION
**Pull request by oktaal**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  - pip install pip-tools
+  - yarn global add @angular/cli
+script:
+  - cookiecutter . --no-input
+  - cd dighum_pro
+  - cat bootstrap.log
+  - cd backend && pip-compile && cd ..
+  - cd functional-tests && pip-compile && cd ..
+  - yarn
+  - yarn django migrate
```
Comment (jgonggrijp) [inline .travis.yml:29]: The previous four steps shouldn't be necessary. Why did you add them?

File: README.md [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
 
 
 ## Before you start
 
 You need to install *at least* the following software:
 
+ - [@angular/cli](https://angular.io/guide/quickstart) for Angular projects.
  - Python >= 3.4, <= 3.6
- - [Cookiecutter][4]
+ - [Cookiecutter][4] (install using pip in a virtualenv using Python 3)
```
Comment (jgonggrijp) [inline README.md:18]: Currently, Python 3.6 only.

File: cookiecutter.json [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
   "description": "{{cookiecutter.project_title}} will humanize all your digits!",
   "author": "UU Digital Humanities Lab",
   "origin": "github:UUDigitalHumanitieslab/{{cookiecutter.slug}}",
   "database_name": "{{cookiecutter.slug}}",
   "database_user": "{{cookiecutter.slug}}",
   "database_password": "{{cookiecutter.slug}}",
   "localizations": "en:english,fr:french",
+  "frontend": ["angular", "backbone"],
+  "frontend_port": "4200",
+  "backend_port": "8000",
```
Comment (jgonggrijp) [inline cookiecutter.json:14]: These settings and the `app_prefix` haven't been documented yet in the README.

File: {{cookiecutter.slug}}/.travis.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    - frontend/node_modules
+addons:
+  chrome: stable
+before_install:
+  - nvm install 10
+  - curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
+  - echo "deb http://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
+  - sudo apt-get update -qq
+  - sudo apt-get install -y -qq yarn
+  - google-chrome-stable --headless --disable-gpu --remote-debugging-port=9222 http://localhost &
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/.travis.yml:20]: In this case (CI-testing the generated project), manually installing yarn shouldn't be necessary because [yarn is enabled by default if a yarn.lock is found](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#yarn-is-supported).

File: {{cookiecutter.slug}}/.travis.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+install:
+  - yarn global add chromedriver geckodriver webdriver
+script:
+  - source .env/bin/activate
+  - yarn
+  - yarn django migrate
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/.travis.yml:27]: The above two steps can be replaced with `python bootstrap.py` and then activating the virtualenv after that. The `bootstrap.py` doesn't have a `--no-input` option, but you can `echo $'\n\n\n\n\n' | python` with enough newlines to pick the default for each question.

File: {{cookiecutter.slug}}/bootstrap.py [hunk 6]
```diff
@@ -163,6 +167,98 @@ def make_create_db_command(psql_cmd):
         psql_cmd + ' -f ' + op.join('backend', 'create_db.sql'),
     )
 
+def merge_json(target, source):
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/bootstrap.py:170]: I think this function and the next two (`create_angular_localization_settings` and `activate_frontend`) belong in `/hooks/post_gen_project.py`, because you only need them when generating a new project, not when cloning an existing project.

File: {{cookiecutter.slug}}/bootstrap.py [hunk 7]
```diff
@@ -163,6 +167,98 @@ def make_create_db_command(psql_cmd):
...
+    return target
+
+def create_angular_localization_settings():
+    with open('frontend/angular.json', 'r') as file:
+        data = json.load(file)
+    project = '{{cookiecutter.slug}}'.replace('_', '-')
+    for lang in '{{cookiecutter.localizations}}'.split(','):
+        [code, lang_name] = lang.split(':')
+        production = merge_json({}, data['projects'][project]['architect']['build']['configurations']['production'])
+        production['outputPath'] = f'dist/{code}'
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/bootstrap.py:188]: This syntax doesn't work in python <3.6.

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  - yarn test-back && yarn test-front
```
Comment (jgonggrijp) [inline .travis.yml]: To make `yarn test-func` work, you need to run `yarn static-p` in advance, then run `yarn serve-p` in the background, then wait about a second for both servers to be ready and finally kick off `yarn test-func`. You might need to manually kill the background task in order for Travis to consider the build complete.

File: {{cookiecutter.slug}}/bootstrap.py [hunk 9]
```diff
@@ -163,6 +167,98 @@ def make_create_db_command(psql_cmd):
...
+
+def activate_frontend():
+    framework = '{{cookiecutter.frontend}}'
+    os.rename('package.{{cookiecutter.frontend}}.json', 'package.json')
+
+    if framework == 'backbone':
+        os.rename('frontend.backbone', 'frontend')
+    elif framework == 'angular':
+        project_name = '{{cookiecutter.slug}}'.replace('_', '-')
+        Command(
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/bootstrap.py:217]: I recommend creating these commands as global objects (i.e., in module scope) and just invoking them here. These objects have a `__print__` method and the invocation (`__call__`) returns a boolean which indicates whether the command was successful, so you can make one command conditional on the success of another and you can also display failed commands to the user so they can run them by themselves. See the `main` function for example usage.

File: {{cookiecutter.slug}}/bootstrap.py [hunk 10]
```diff
@@ -163,6 +167,98 @@ def make_create_db_command(psql_cmd):
...
+            data = json.load(file)
+        with open('frontend/package.json', 'w') as file:
+            merge_json(data, overwrite)
+            json.dump(data, file, indent=2)
+        os.remove('frontend/package.overwrite.json')
+    # remove other frameworks
+    for path in glob.glob("frontend.*"):
+        shutil.rmtree(path)
+    for path in glob.glob("package.*.json"):
+        os.remove(path)
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/bootstrap.py:269]: I recommend returning a boolean that indicates whether `activate_frontend` was successful overall, because you may want to make some steps in the `main` conditional on this.

File: {{cookiecutter.slug}}/frontend.angular/proxy.conf.json [hunk 11]
```diff
@@ -0,0 +1,6 @@
+{
+    "/api": {
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/frontend.angular/proxy.conf.json:2]: This shouldn't be just `/api`, but everything except for the static assets that aren't the index.html. See the Backbone frontend proxy for comparison.

File: {{cookiecutter.slug}}/package.backbone.json [hunk 12]
```diff
@@ -22,19 +22,19 @@
     "test-back": "yarn back pytest",
     "test-func": "yarn func pytest",
     "test": "yarn test-back && yarn test-front && yarn test-func",
-    "watch-front": "yarn gulp watch --config ../config.json",
+    "watch-front": "yarn gulp watch --config config.root.json",
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/package.backbone.json:25]: This should be changed back, and the config file in question should be moved to the old location by the post-generation hook.

File: {{cookiecutter.slug}}/package.backbone.json [hunk 13]
```diff
@@ -22,19 +22,19 @@
...
     "start-front": "yarn gulp clean && yarn watch-front",
     "start-back": "yarn django runserver --settings settings --pythonpath ..",
     "livereload-back": "yarn django livereload --settings settings --pythonpath ..",
     "watch-test-back": "yarn test-back --looponfail",
     "watch-back": "yarn livereload-back & yarn watch-test-back",
     "start": "yarn start-front & yarn start-back & yarn watch-back",
     "dist": "yarn gulp dist",
-    "dist-p": "yarn dist --production --config ../config.json",
+    "dist-p": "yarn dist --production --config config.root.json",
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/package.backbone.json:33]: Same here

File: {{cookiecutter.slug}}/package.backbone.json [hunk 14]
```diff
@@ -22,19 +22,19 @@
...
     "collectstatic-p": "yarn django collectstatic -v 0 --settings production --pythonpath ..",
     "clean-static": "rm -rf static || rd /S /Q static",
     "static-p": "yarn gulp clean && yarn dist-p && yarn clean-static && yarn collectstatic-p",
-    "serve-front-p": "yarn gulp serve --proxy ../proxy.json --root ..",
+    "serve-front-p": "yarn gulp serve --proxy proxy.json --root ..",
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/package.backbone.json]: Same here.

File: {{cookiecutter.slug}}/package.angular.json [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        "install-back": "yarn back -- pip install -r requirements.txt",
+        "install-func": "yarn func -- pip install -r requirements.txt",
+        "postinstall": "yarn fyarn && yarn install-back && yarn install-func",
+        "django": "yarn back python manage.py",
+        "startapp": "yarn back django-admin startapp --settings itf_manager.settings --pythonpath .",
+        "test-front": "yarn front yarn test-once",
+        "test-back": "yarn back pytest",
+        "test-func": "yarn func pytest",
+        "test": "yarn test-back && yarn test-front && yarn test-func",
+        "watch-front": "yarn front yarn serve --proxy-config proxy.conf.json",
```
Comment (jgonggrijp) [inline {{cookiecutter.slug}}/package.angular.json]: Proxy configuration belongs in the project root.

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
I can tell you put serious effort into properly integrating the Angular frontend with the existing infrastructure. Well done on that. Also kudos for adding two new features along the way:

 - configurable frontend/backend ports;
 - travis configuration, both for the scaffold and for the generated project.

You spent extra effort to keep Backbone as an alternative choice. I value this on personal grounds, but I also think this is a good design choice for future maintenance reasons:

 - Although the current approach with renaming directories and moving files adds some complexity, this complexity will even out again when we modularize the scaffolding tool (so you can e.g. also choose to generate only a frontend, or a backend and functional tests but no frontend). This is something I've been thinking about since I started the scaffold, anyway; I'll create an issue for it after completing this review.
 - The Angular version of the scaffold can benefit from backend/functest/integration improvements in existing Backbone projects and vice versa.

I tried the cookiecutter. Of course, some things don't work yet; see the main issues below.

 - [x] The generated `bootstrap.py` fails with a syntax error (on line 188, `f'dist/{code}'`) if cookiecutter is installed in a virtualenv with Python <3.6. The same gotcha applies when somebody who clones the project manually runs the bootstrap script. This should either be document or changed to be compatible with Python >=3.4.
 - [x] The generated README in the project root still reflects the Backbone scenario even if you choose Angular.
 - [x] The Angular scenario has an inversion of control compared to the Backbone scenario: the frontend server is in charge of serving the index page instead of the backend. This breaks the provision to ensure that the CSRF cookie is present from first page load. Either move responsibility for the index page to the backend again (see next two points), or add a backend API and a frontend request to get the CSRF cookie after the fact. I'd prefer the first option because it saves a request, because it provides better isolation and because it keeps the divergence between the Angular and the Backbone version to a minimum. I also think this will be less work for you overall.
 - [x] The index on `localhost:8000/` is broken (Django cannot find the index.html). The `settings.py` in the project root needs to be updated for the Angular scenario in order to fix this; see the `STATICFILES_DIRS` key.
 - [x] To undo the inversion of control, the frontend proxy should forward *all* requests to the backend, except for static files that are not the index.html. We can have a look at this together if you wish; see also the proxy configuration of the Backbone version. Please note that you only need the frontend server in production mode, because in development mode the backend server can do everything (including livereload).
 - [x] `localhost:4200/api/` leads to the browsable API, but the stylesheets cannot be loaded.
 - [x] `localhost:4200/admin/` loops back to the index page. The `test_{{cookiecutter.slug}}_admin` functional test fails for this reason.
 - [x] If I run `yarn start-p` and open `localhost:4200`, the developer console is still telling me that Angular is running in development mode.
 - [x] The `test_{{cookiecutter.slug}}_frontend` functional test assumes that the `{{cookiecutter.project_title}}` is in the browser title. This needs to change for the Angular scenario.
 - [x] The Angular frontend still contains the default e2e suite. I recommend removing this, since it is redundant with the `/functional-tests` sub-project. The latter is in pytest, which we already use in most of our Python applications, and which is also easier to use and framework-agnostic. Moreover, it already contains provisions for testing a live application on a deployment server.
 - [ ] The Django backend runs on port 8000 regardless of the backend port setting.
 - [ ] The Backbone frontend runs on port 8080 regardless of the frontend port setting.
 - [x] "Magic" configuration which integrates knowledge about both the frontend and the backend, like the proxy settings and the file that you renamed to config.root.json in the Backbone variant, belongs in the root directory, not in the frontend directory. It is OK to keep these files in the `frontend.angular` and `frontend.backbone` directories before generation, but they should be moved to the root by the post generation hook.

Details, comments, tips etcetera below. There is a lot to consider so I probably missed some things, too.

I think you're on the right track. Keep it up and good luck!

File: .travis.yml (hunk 1 above)
Comment (oktaal) [reply .travis.yml:29]: This is a work-around for the created virtualenv being in Python 2 for some reason:

> /home/travis/build/UUDigitalHumanitieslab/cookiecutter-webapp-deluxe/dighum_pro/.env/local/lib/python2.7/site-packages/pip/_vendor/urllib3/util/ssl_.py:354: SNIMissingWarning: An HTTPS request has been made, but the SNI (Server Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
>   SNIMissingWarning

https://travis-ci.com/UUDigitalHumanitieslab/cookiecutter-webapp-deluxe/builds/108120856

The same also happened on my Linux machine (using a Python 3 virtualenv). So I added this workaround to use the virtualenv created by Travis itself which is using Python 3.

File: .travis.yml (hunk 8 above)
Comment (oktaal) [reply .travis.yml]: I already tried this (well apart from `yarn static-p`), it doesn't work because the web drivers fail:
> selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable needs to be in PATH. Please see https://sites.google.com/a/chromium.org/chromedriver/home

https://travis-ci.com/UUDigitalHumanitieslab/cookiecutter-webapp-deluxe/jobs/192507563

File: .travis.yml (hunk 1 above)
Comment (jgonggrijp) [reply .travis.yml:29]: This is probably the builtin virtualenv of the OS image. `pip install virtualenv` before starting the cookiecutter might resolve this.

File: .travis.yml (hunk 8 above)
Comment (jgonggrijp) [reply .travis.yml]: You might just need to `yarn global add chromedriver geckodriver` before running `yarn test`.

By the way, if you do `yarn test`, the unittests run first so you don't need to fiddle with a 1-second timeout. You still need to get hold of the pid of `yarn serve-p` so you can stop it, though.
